language/move-lang/src/unit_test/plan_builder.rs

```diff
@@ -0,0 +1,325 @@
+// Copyright (c) The Diem Core Contributors
```

Comment:
It does feel a little bit awkward to me that this is being implemented in the compiler crate instead of the unit test crate. 

I do understand that what I suggested may not be possible today, but still it would be a good practice for us to think about:
- What is the role of the move compiler crate? Should we limit it to just parsing and compiling?
- How in general are we going to support tools that processes parsed or partially compiled Move code? What kind of APIs do we want to expose and what are the stability guarantees?

Reply:
I think we could pull this out into the `move-unit-test` crate if we wanted to. I may be off-base here, but (IMO) that's probably not something we want to do most likely. My main reason here is that in many ways, I view unit testing as a part of the source language, and because of this it should be a part of the compiler and compilation pipeline. This then leads to your first question: 

> What is the role of the move compiler crate? Should we limit it to just parsing and compiling?

In my view a test plan is a compilation artifact, just like a source map or a compiled module. What we then do with these compilation artifacts (disassembler, vm, unit test runner, etc.) is up to the consumers of them, but they're not responsible for the generation of those artifacts. 

One other reason I think this should most likely be a part of the compilation pipeline is that since the compiler needs to be aware of the testing flag and has testing "mode", if we wanted to split the test plan building into a different crate, we would still need to have well-formedness of the unit test constructs done in the compiler. If that's the case, it feels a bit odd to me to separate these two things -- it seems more sensible to have the thing that is responsible for checking and verifying the unit tests to also be the one constructing the executable artifact.

Reply:
I think this can be reviwed a bit more once/if we have some sort of attribute schema system. If that happens, you can split off the verification of the attributes from the actual test plan building. And move that over to `move-unit-test`